WRF starting and ending dims for MPI decomposition

davegill opened the pull request:
Namelist:
```
 &domains
 e_we                                = 74,    112,   94,
 e_sn                                = 61,    97,    91,
 /
```

1. One processor:
```
> wrf.exe ; tail rsl.out.0000
 starting wrf task            0  of            1
DYNAMICS OPTION: Eulerian Mass Coordinate
   alloc_space_field: domain            1 ,             144490964  bytes allocated
  med_initialdata_input: calling input_input
   Input data is acceptable to use: wrfinput_d01
Timing for processing wrfinput file (stream 0) for domain        1:    0.14616 elapsed seconds
              1             74              1             61
-------------- FATAL CALLED ---------------
FATAL CALLED FROM FILE:  <stdin>  LINE:     224
just stop it
-------------------------------------------
```

2. Two processors
```
> mpirun -np 2 wrf.exe ; tail rsl.out.0000
 starting wrf task            0  of            2
 starting wrf task            1  of            2
   alloc_space_field: domain            1 ,              87213260  bytes allocated
  med_initialdata_input: calling input_input
   Input data is acceptable to use: wrfinput_d01
Timing for processing wrfinput file (stream 0) for domain        1:    0.17207 elapsed seconds
              1             74              1             30
              1             74             31             61
-------------- FATAL CALLED ---------------
FATAL CALLED FROM FILE:  <stdin>  LINE:     224
just stop it
-------------------------------------------
```

3. Three processors
```
> mpirun -np 3 wrf.exe ; tail rsl.out.0000
 starting wrf task            0  of            3
 starting wrf task            1  of            3
 starting wrf task            2  of            3
  med_initialdata_input: calling input_input
   Input data is acceptable to use: wrfinput_d01
Timing for processing wrfinput file (stream 0) for domain        1:    0.17245 elapsed seconds
              1             74              1             20
              1             74             21             40
              1             74             41             61
-------------- FATAL CALLED ---------------
FATAL CALLED FROM FILE:  <stdin>  LINE:     224
just stop it
-------------------------------------------
```

4. Four processors
```
> mpirun -np 4 wrf.exe ; tail rsl.out.0000
 starting wrf task            0  of            4
 starting wrf task            2  of            4
 starting wrf task            3  of            4
 starting wrf task            1  of            4
   Input data is acceptable to use: wrfinput_d01
Timing for processing wrfinput file (stream 0) for domain        1:    0.18871 elapsed seconds
              1             37              1             30
             38             74              1             30
              1             37             31             61
             38             74             31             61
-------------- FATAL CALLED ---------------
FATAL CALLED FROM FILE:  <stdin>  LINE:     224
just stop it
-------------------------------------------
```

davegill committed:
WRF starting and ending dims for MPI decomposition

1. One processor:
```
> wrf.exe ; tail rsl.out.0000
 starting wrf task            0  of            1
DYNAMICS OPTION: Eulerian Mass Coordinate
   alloc_space_field: domain            1 ,             144490964  bytes allocated
  med_initialdata_input: calling input_input
   Input data is acceptable to use: wrfinput_d01
Timing for processing wrfinput file (stream 0) for domain        1:    0.14616 elapsed seconds
              1             74              1             61
-------------- FATAL CALLED ---------------
FATAL CALLED FROM FILE:  <stdin>  LINE:     224
just stop it
-------------------------------------------
```

2. Two processors
```
> mpirun -np 2 wrf.exe ; tail rsl.out.0000
 starting wrf task            0  of            2
 starting wrf task            1  of            2
   alloc_space_field: domain            1 ,              87213260  bytes allocated
  med_initialdata_input: calling input_input
   Input data is acceptable to use: wrfinput_d01
Timing for processing wrfinput file (stream 0) for domain        1:    0.17207 elapsed seconds
              1             74              1             30
              1             74             31             61
-------------- FATAL CALLED ---------------
FATAL CALLED FROM FILE:  <stdin>  LINE:     224
just stop it
-------------------------------------------
```

3. Three processors
```
> mpirun -np 3 wrf.exe ; tail rsl.out.0000
 starting wrf task            0  of            3
 starting wrf task            1  of            3
 starting wrf task            2  of            3
  med_initialdata_input: calling input_input
   Input data is acceptable to use: wrfinput_d01
Timing for processing wrfinput file (stream 0) for domain        1:    0.17245 elapsed seconds
              1             74              1             20
              1             74             21             40
              1             74             41             61
-------------- FATAL CALLED ---------------
FATAL CALLED FROM FILE:  <stdin>  LINE:     224
just stop it
-------------------------------------------
```

4. Four processors
```
> mpirun -np 4 wrf.exe ; tail rsl.out.0000
 starting wrf task            0  of            4
 starting wrf task            2  of            4
 starting wrf task            3  of            4
 starting wrf task            1  of            4
   Input data is acceptable to use: wrfinput_d01
Timing for processing wrfinput file (stream 0) for domain        1:    0.18871 elapsed seconds
              1             37              1             30
             38             74              1             30
              1             37             31             61
             38             74             31             61
-------------- FATAL CALLED ---------------
FATAL CALLED FROM FILE:  <stdin>  LINE:     224
just stop it
-------------------------------------------
```

billsacks commented:
@davegill thanks a lot for putting this together. I'm not clear on how exactly to use this information yet, but this seems like a good starting point.

We need two things:

(1) A 1-d indexing of all of the grid cells in the WRF domain. That is, each grid cell should have a unique index 1..n (where n = total number of grid cells), and this indexing needs to be independent of the decomposition. This indexing will need to match the indexing on the ESMF mesh file that is provided to LILAC. (We can help with creation of this mesh file, but I bring up this last point to emphasize that it will probably be easiest if this indexing is in some rational ordering - e.g., South to North, West to East.)

(2) Each processor needs a vector of size n_local (i.e., number of grid cells owned by this processor), mapping local indices to global indices. Let's say a given processor owns grid cells 2, 5, 7 and 11 in global index space. Then it will have the vector [2,5,7,11].

I'm not clear on how much of this information already exists in WRF and how much we'll need to create specially for this purpose. This – along with any guidance you can provide regarding how to create the necessary numberings – is what we're hoping you can help with.

davegill commented:
@billsacks 
Bill,
Something like this?
```



     =================     =================
     |57 |58 |59 |60 |     |61 |62 |63 |64 |
     -----------------     -----------------
     |49 |50 |51 |52 |     |53 |54 |55 |56 |
     -----------------     -----------------
     |41 |42 |43 |44 |     |45 |46 |47 |48 |
     -----------------     -----------------
     |33 |34 |35 |36 |     |37 |38 |39 |40 |
     =================     =================
        MPI TASK 2            MPI TASK 3
    

     =================     =================
     |25 |26 |27 |28 |     |29 |30 |31 |32 |
     -----------------     -----------------
     |17 |18 |19 |20 |     |21 |22 |23 |24 |
     -----------------     -----------------
     | 9 |10 |11 |12 |     |13 |14 |15 |16 |
     -----------------     -----------------
     | 1 | 2 | 3 | 4 |     | 5 | 6 | 7 | 8 |
     =================     =================
        MPI TASK 0            MPI TASK 1



MPI TASK 0 owns:  1,  2,  3,  4,  9, 10, 11, 12, 17, 18, 19, 20, 25, 26, 27, 28
MPI TASK 1 owns:  5,  6,  7,  8, 13, 14, 15, 16, 21, 22, 23, 24, 29, 30, 31, 32
MPI TASK 2 owns: 33, 34, 35, 36, 41, 42, 43, 44, 49, 50, 51, 52, 57, 58, 59, 60
MPI TASK 3 owns: 37, 38, 39, 40, 45, 46, 47, 48, 53, 54, 55, 56, 61, 62, 63, 64
```

billsacks commented:
@davegill yes that is exactly what we need – thank you for the great example / picture!

davegill committed:
MPI tiling info

Info is now dumped to fort.10

Changes to be committed:
modified:   dyn_em/start_em.F

davegill commented:
@billsacks 
Bill,
OK, the code now generates a text-based fort.10 file that looks like the following:
```
MPI RANK =      0 OF      4 Processes
       1     (   1,   1)
       2     (   2,   1)
       3     (   3,   1)

    2152     (  35,  30)
    2153     (  36,  30)
    2154     (  37,  30)


MPI RANK =      1 OF      4 Processes
      38     (  38,   1)
      39     (  39,   1)
      40     (  40,   1)

    2188     (  71,  30)
    2189     (  72,  30)
    2190     (  73,  30)


MPI RANK =      2 OF      4 Processes
    2191     (   1,  31)
    2192     (   2,  31)
    2193     (   3,  31)

    4342     (  35,  60)
    4343     (  36,  60)
    4344     (  37,  60)


MPI RANK =      3 OF      4 Processes
    2228     (  38,  31)
    2229     (  39,  31)
    2230     (  40,  31)

    4378     (  71,  60) 
    4379     (  72,  60) 
    4380     (  73,  60)
```

The only reason to dump it out is so that you can see what the values look like. After the call to `feed_the_machine`, the following arrays are filled:
```
owned_per_MPI_task_cells
owned_per_MPI_task_i
owned_per_MPI_task_j
```
These are each dimensioned by (number of cells per MPI rank, MPI rank).

billsacks commented:
Fantastic - thank you @davegill ! I have some questions about this that we can discuss this afternoon, but it looks like this will basically give us what we need.

billsacks commented:
@davegill - @negin513 and I are starting to work on this. Can you clarify for us your preference on where this new code ends up. Options that would all work about equally for us are:

(1) We keep all the new code for calculating `owned_per_MPI_task_cells` in the new lilac-specific module

(2) We put the new subroutine to compute `owned_per_MPI_task_cells` in start_em.F and call it from the lilac-specific module

(3) We put the new subroutine to compute `owned_per_MPI_task_cells` in start_em.F and call it from `start_domain_em` (similar to what you have done), saving the result somewhere so that the lilac-specific module and anyone else can reference this.

I think the best choice mainly depends on how useful you think this will be for other applications, so we're open to whatever you prefer here.

davegill commented:
@billsacks @negin513
Bill and Negin,
1. The computations that are required are quite minimal.
2. Likely, there is a longer-term plan to have LILAC hook up to other models.

I would lean towards option 2.

Make this a new module that is inside of the WRF system. Have the Fortran interface of LILAC *use* that new WRF module. Make the call somewhat generic, so that you can essentially make every other model also have the same interface within their own source code.
